simulation/Chart.ts

```diff
@@ -0,0 +1,107 @@
+import { ChartJSNodeCanvas, ChartCallback } from 'chartjs-node-canvas';
+import { ChartConfiguration } from 'chart.js';
+import * as fs from 'fs/promises';
+import * as nodepath from 'path';
+
+export interface SimulationResults {
+  [label: string]: {
+    utilization: number,
+    totalSupply: number,
+  }
+};
+
+function createData(simResults: SimulationResults) {
+  const colors = [
+    'rgba(255, 99, 132, 0.85)',
+    'rgba(54, 162, 235, 0.85)',
+    'rgba(255, 206, 86, 0.85)',
+    'rgba(75, 192, 192, 0.85)',
+    'rgba(153, 102, 255, 0.85)',
+    'rgba(255, 159, 64, 0.85)'
+  ];
+  const datasets = Object.keys(simResults).map((label, i) => {
+    return {
+      label,
+      data: [{
+        x: simResults[label].totalSupply,
+        y: simResults[label].utilization
+      }],
+      pointRadius: 5,
+      backgroundColor: [colors[i]]
+    }
+  });
+  return { datasets };
+}
+
+export async function createChart(simResults: SimulationResults, title?: string): Promise<void> {
+  const width = 800;
+  const height = 600;
+  const data = createData(simResults);
+
+  const configuration: ChartConfiguration = {
+    type: 'scatter',
+    data,
+    options: {
+      plugins: {
+        title: {
+          display: title !== undefined,
+          text: title ? title : ''
+        },
+        legend: {
+          position: 'right'
+        }
+      },
+      scales: {
+        x: {
+          type: 'linear',
+          position: 'bottom',
+          title: {
+            display: true,
+            text: 'Total Supply (M)'
+          }
+        },
+        y: {
+          type: 'linear',
+          min: 0.45,
+          max: 0.85,
+          title: {
+            display: true,
+            text: 'Utilization'
+          }
+        }
+      }
+    },
+    plugins: [{
+      id: 'background-colour',
+      beforeDraw: (chart) => {
+        const ctx = chart.ctx;
+        ctx.save();
+        ctx.fillStyle = 'white';
+        ctx.fillRect(0, 0, width, height);
+        ctx.restore();
+      },
+    }]
+  };
+  const chartCallback: ChartCallback = (ChartJS) => {
+    ChartJS.defaults.responsive = true;
+    ChartJS.defaults.maintainAspectRatio = false;
+  };
+  const chartJSNodeCanvas = new ChartJSNodeCanvas({ width, height, chartCallback });
+  const buffer = await chartJSNodeCanvas.renderToBuffer(configuration);
+  const fileName = title ? `${title}.png` : 'chart.png';
+  const path = nodepath.join(process.cwd(), 'simulation', 'charts', fileName);
+  const dir = nodepath.dirname(path);
+  if (!(await fileExists(dir))) {
+    await fs.mkdir(dir, { recursive: true });
+  }
+  await fs.writeFile(path, buffer, 'base64');
+}
+
+async function fileExists(path: string): Promise<boolean> {
+  try {
+    await fs.stat(path);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
```

Comment:
Very cool! Want to upload a png? Not sure about introducing this dep though, its failing to install for me :(

Reply:
Maybe add to `optionalDependencies`?

Reply:
Sounds good. I can make the dependencies optional.

Sure, I'll upload the pngs (they are excluded by `.gitignore` right now). It's still a WIP because I want to also visualize how profitable (or at least if profitible) each of these points are.

Reply:
You should be able to view the charts now.

Reply:
Hm I can't run yarn anymore, which doesn't seem to be related to this branch necessarily (something to do with `hardhat-cover` actually, so I'll debug that separately). I still can't install these optionalDependencies (out of the box at least), but you are saying I can view them bc you added them to the repo? I guess that's cool (and thanks), but I was just thinking upload them to the PR. You think we should store them in the repo? Might not be a bad idea if not everyone can run these sims easily but also large-ish binary files and could get outdated

Reply:
Oh yeah that's smarter 🤦.  I wasn't planning on merging the images, but committed them temporarily so you can view them. Let me just upload them directly to the PR instead.

Reply:
Ok fixed the other issue using `npm i`, but still can't compile with these optional dependencies since the typescript compilation fails on `import { ChartConfiguration } from 'chart.js';`. So hesitant to merge, maybe we can find a different way of plotting that doesn't require native deps to be setup correctly so this can be portable